### PR TITLE
Test Fixes to Run Xcode 15 in CI 

### PR DESCRIPTION
**Pull request by jostnes**
### Description
This is the "cleaned up" version of @mokagio's PR https://github.com/wordpress-mobile/WordPress-iOS/pull/21921, with only the final working commit and the many experimentation commits from the original PR omitted (see original PR for more description about the changes)

The branch was tested on [WPiOS-macv2-test](https://buildkite.com/automattic/wpios-macv2-test/builds?branch=ui-test-fixes-ios17.2) with the UI test steps passing on every build. There are still some flaky tests, but automated retries currently handle those, and it would be better to handle them separately if they become too flaky. The current run time for UI test is still within 30 minutes (same as before).

The UI test fixes that were made:
1. Somehow stats test started failing/became flakier with this change. I updated the stats test and mocks to be more dynamic; the year is no longer hard coded to 2019 but follows the current year. Flakiness is still seen when the screen is not updated to use the values presented in the mocks; I couldn't reproduce this on a real device, which is why this could be a mock-related issue. 
2. On reply notification, another flakiness is seen where the tap does not work (based on screenshots, the tap did happen, but the action that should have happened after the tap did not. Also, it is not reproducible on a real device). I renamed `tapUntilCondition` to `waitForElementAndTap` and updated it to cover more conditions to make it more robust. The change works for now.
3. To workaround the simulator bug where we can't turn off Save Password from settings, `app.dismissSavePasswordPrompt()` was readded after password screen.

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist: Not a UI PR

### CODE REVIEW

File: WordPress/UITestsFoundation/Globals.swift
Comment (tiagomar) [inline WordPress/UITestsFoundation/Globals.swift]: I'm not sure `waitForElementAndTap `  describes exactly what the function does? The function doesn't actually wait for the `tapElement ` before tapping on it. 

WDYT about `tapUntil(_ tapElement: XCUIElement, conditionOn conditionElement: XCUIElement, condition: String, errorMessage: String, maxRetries: Int = 10)`?

**Comment by tiagomar** [comment]
The `executeWithRetries(disableAutoFillPasswords)` in `testBundleWillStart` from `TestObserver` is still running.[ Is the idea still to skip it for now](https://github.com/wordpress-mobile/WordPress-iOS/pull/21921#discussion_r1429653310))?

**Review by tiagomar — CHANGES_REQUESTED** [review]
Thanks for working on this, @jostnes ! 🙇 

I've run all tests locally on Xcode 15.1 and all tests passed on iPhone and iPad. I just left a couple comments before approving this PR.

File: WordPress/UITestsFoundation/Globals.swift
Comment (mokagio) [reply WordPress/UITestsFoundation/Globals.swift]: @tiagomar 's observation resonates with me. As a matter of fact, I opened https://github.com/wordpress-mobile/WordPress-iOS/pull/22322 to address it (with an attempt to improve upon it).

Having said that, I don't see this as a blocker to merge given the benefit that merging this PR to add Xcode 15.1 / iOS 17.2 support would bring to the dev teams.

**Review by mokagio — APPROVED** [review]
@jostnes @tiagomar I'd love to see this merged sooner rather than later, even if it means not addressing the naming comment straightaway and instead follow up later. Rationale: the sooner we officially support Xcode 15.1 and iOS 17+ in CI, the better.

**Comment by staskus** [comment]
Note: I have widget API compliance for Xcode 15 + iOS 17 lined up (https://github.com/wordpress-mobile/WordPress-iOS/pull/21705). So it's not a blocker if you notice anything wrong related to widgets.

File: WordPress/UITestsFoundation/Globals.swift
Comment (tiagomar) [reply WordPress/UITestsFoundation/Globals.swift]: I agree, @mokagio, it makes sense to move on with this PR and address the naming in another one. 👍

**Review by tiagomar — APPROVED** [review]
Approving this PR now as the naming issue will be addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/22322.

**Comment by jostnes** [comment]
There's a unit test failing in the PR. I'll update with the latest trunk again to see if it solves the issue and to be able to merge this.

**Comment by tiagomar** [comment]
> There's a unit test failing in the PR. I'll update with the latest trunk again to see if it solves the issue and to be able to merge this.

Yeah, it has been taken care of  in https://github.com/wordpress-mobile/WordPress-iOS/pull/22322/commits/7365c10817533903d09d145e00c7c84bd51c5ab7. https://github.com/wordpress-mobile/WordPress-iOS/pull/22322 is ready to be merged into this branch.

**Comment by jostnes** [comment]
@mokagio the build is good on the [iOS 17+ pipeline in CI](https://buildkite.com/automattic/wpios-macv2-test/builds/42) but some steps are failing on [the pipeline checks linked to this PR ](https://buildkite.com/automattic/wordpress-ios/builds/19458#018cd277-66d4-4729-931e-730b13469127)because of `Unable to find remote image: xcode-15.1` (which i guess is expected)

should the GitHub checks be updated for this? at least one of the failing step is marked as required so we can't merge this yet.

**Comment by mokagio** [comment]
@jostnes yes sorry, I should have left a note that before being able to merge this we need to move WordPress-iOS on Buildkite to the _Apple Silicon [cluster](https://buildkite.com/docs/clusters/overview)_.

I have a PR for it that I'll mention internally.

**Comment by guarani** [comment]
> Note: I have widget API compliance for Xcode 15 + iOS 17 lined up (#21705). So it's not a blocker if you notice anything wrong related to widgets.

👋 Will the 24.0 beta be built using Xcode 15? If so, I think #21705 needs to be merged before Monday to avoid widgets breaking in the beta builds. Apologies, it's been waiting for my review for a while now, I'll get to it ASAP.

**Comment by mokagio** [comment]
I just merged the PR to move this CI stack to Apple Silicon and support `image: xcode-15.1`. 

I triggered a new build which should run on that cluster and be successful in the same way the past builds on the test pipeline were.

https://buildkite.com/automattic/wordpress-ios/builds/19537

**Pull request merged**
